### PR TITLE
Default revision implementation

### PR DESCRIPTION
**Pull request by Monkeyanator**
Implementation for `defaultRevision`, dependent on [https://github.com/istio/api/pull/1806](https://github.com/istio/api/pull/1806). Creates an `istiod` service for the default revision and repoints it when another revision claims to be the default (rather than the approach taken [here](https://github.com/istio/istio/pull/30286)) @costinm @howardjohn @zerobfd 

[ ] Configuration Infrastructure
[ ] Docs
[x] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[ ] Does not have any changes that may affect Istio users.

### CODE REVIEW

**Review by howardjohn — APPROVED** [review]
Is the plan to add default injector in another PR?

File: releasenotes/notes/default-revision.yaml
Comment (shamsher31) [inline releasenotes/notes/default-revision.yaml:8]:
```suggestion
    **Added**  the `defaultRevision` option to the `IstioOperator`. Makes it possible to specify which Istio control plane
```

Comment (Monkeyanator) [reply releasenotes/notes/default-revision.yaml]: Done, thanks

**Comment by Monkeyanator** [comment]
> Is the plan to add default injector in another PR?

Yes but might include it here and re-request review depending on how long the API change takes to get merged

**Comment by Monkeyanator** [comment]
@howardjohn added the default injector here, changes are pretty substantial so may be worth taking another look (adds separate `MutatingWebhookConfiguration`) 

cc @su225 since this may require changes to revision and tag istioctl commands

**Review by howardjohn — CHANGES_REQUESTED** [review]

File: manifests/charts/istiod-remote/templates/mutatingwebhook.yaml
Comment (Monkeyanator) [reply manifests/charts/istiod-remote/templates/mutatingwebhook.yaml]: Ah yep, should be "true", thanks

File: manifests/charts/istio-control/istio-discovery/templates/mutatingwebhook.yaml
Comment (Monkeyanator) [reply manifests/charts/istio-control/istio-discovery/templates/mutatingwebhook.yaml:57]: So when doing `istioctl install`, the default injection logic will be in `istio-sidecar-injector-default`, and it will get overwritten when doing `istioctl install --set revision=foo --defaultRevision`.

This is somewhat confusing since the default install will result in two `MutatingWebhookConfiguration`s, `istio-sidecar-injector` with no webhooks (since all webhooks in that config are per-revision), and `istio-sidecar-injector-default` with 2

File: istioctl/cmd/tag.go
Comment (Monkeyanator) [reply istioctl/cmd/tag.go:41]: Added a comment. Pretty much it'll be useful for commands to be able to find resources relating to the default revision

File: manifests/charts/istio-control/istio-discovery/templates/mutatingwebhook.yaml
Comment (howardjohn) [reply manifests/charts/istio-control/istio-discovery/templates/mutatingwebhook.yaml:57]: Oh, I see. I was expecting the "default" one to be `istio-sidecar-injector` and the per revision one to be `istio-sidecar-injector-<revision>`. This seems a bit confusing to me and I am not sure it actually works.
```
$ ~/istio/istio-1.9.0/bin/istioctl install -y
$ grun ./istioctl/cmd/istioctl install -y -d manifests --set revision=bar --set defaultRevision=true
$ kgmwh -L istio.io/rev -L istio.io/default
NAME                             WEBHOOKS   AGE    REV       DEFAULT
istio-sidecar-injector           1          112s   default
istio-sidecar-injector-bar       2          52s    bar
istio-sidecar-injector-default   2          100s   bar       true
```

both istio-sidecar-injector and istio-sidecar-injector-default are matching istio-injection=enabled

Comment (Monkeyanator) [reply manifests/charts/istio-control/istio-discovery/templates/mutatingwebhook.yaml]: Ah I see now, you're right, we definitely need to overwrite the `istio-sidecar-injector` webhook from the old default install.

Comment (Monkeyanator) [reply manifests/charts/istio-control/istio-discovery/templates/mutatingwebhook.yaml]: Looks like `istioctl install` actually tries to merge resources with the same name rather than overwriting, which breaks the logic here for both overwriting the `istiod` service and the `istio-sidecar-injector` webhook (for example: adds labels to service selector rather than overwriting them). Trying to figure out how to get around this

Comment (howardjohn) [reply manifests/charts/istio-control/istio-discovery/templates/mutatingwebhook.yaml]: I have seen a number of issues about this behavior. @richardwxn @ostromart any ideas?

Comment (zerobfd) [reply manifests/charts/istio-control/istio-discovery/templates/mutatingwebhook.yaml]: Is there a way to explicitly patch it with a K8sObjectOverlay? https://github.com/istio/istio/blob/d706ab80df4b1742b299b6c38bf80cb967697efc/pkg/config/labels/instance.go#L97

Comment (Monkeyanator) [reply manifests/charts/istio-control/istio-discovery/templates/mutatingwebhook.yaml]: Even after using the `K8sObjectOverlay` to change the generated config I'm not sure the installer wouldn't just go forward with a merge rather than an overwrite when actually applying it (haven't tried this out though, could be wrong)

Ideally there'd be an annotation we could put on the resource to let the installer know it should overwrite rather than attempt a merge?

File: operator/cmd/mesh/install.go
Comment (howardjohn) [inline operator/cmd/mesh/install.go:161]: nice! one concern though is I don't think this would work with external istiod. For example, I am deploying just remote chart and have `injectionURL` set. The Service will never exist, so now every install will set it as default

**Review by howardjohn — APPROVED** [review]

**Review by shamsher31 — APPROVED** [review]

File: manifests/charts/istiod-remote/values.yaml
Comment (linsun) [inline manifests/charts/istiod-remote/values.yaml:158]: @Monkeyanator could you explain why this is needed if I specify revision=default for config in line 155?

Comment (linsun) [reply manifests/charts/istiod-remote/values.yaml:158]: Also, what if there is multiple defaults?

I think I got it, users can specify revision: 1.9, and set it to default.  Thoughts on configuring it as:
revision: "1.9"
  default: false|true
?

File: manifests/charts/istio-control/istio-discovery/files/gen-istio.yaml
Comment (linsun) [inline manifests/charts/istio-control/istio-discovery/files/gen-istio.yaml:940]: How about `istio.io/rev.default` or `istio.io/rev-default`? It took me a while to associate this `istio.io/default` with the revision configured earlier.

Comment (howardjohn) [reply manifests/charts/istio-control/istio-discovery/files/gen-istio.yaml:940]: I am still not sure why we need this. Previously it was mentioned to distnguish default vs revision items - but we already know this by their name

Comment (Monkeyanator) [reply manifests/charts/istio-control/istio-discovery/files/gen-istio.yaml:940]: Yeah my thinking was that it'd be nice to have a consistent way of getting objects associated with `defaultRevision` across all resource types rather than relying on name but this may be solving for a problem we don't actually have.

Comment (howardjohn) [reply manifests/charts/istio-control/istio-discovery/files/gen-istio.yaml:940]: Yeah I see the idea, and it makes sense, but there is also additional cost/risk of confusion/complexity of introducing a new label. In this case, the cost and the benefit is low, so I don't have a strong opinion either way.

File: manifests/charts/istiod-remote/values.yaml
Comment (Monkeyanator) [reply manifests/charts/istiod-remote/values.yaml:158]: Having multiple `defaultRevision`s will be an issue when using the in-cluster operator since they'll both try and reconcile the `istiod` service to point to themselves. We'll need to make one default and disable the other as atomically as possible to avoid this- this is more a limitation of the API than the impl unfortunately. We actually have this same issue for revision tags and will probably need a declarative API or CRD to do better there.

With other install methods, it'll just work since the base resources get overwritten once on install.

```yaml
revision: "1-9"
  default: false|true
```
looks good to me but I think we should keep it consistent with how it is in the `IstioOperator` API for the time being (which is currently [default_revision](https://github.com/istio/api/blob/35b7ba8ee9c54aed440d5e5b731b5d332b99b5d5/operator/v1alpha1/operator.proto#L98)at same level as `revision`).

**Comment by Monkeyanator** [comment]
@costinm could you PTAL when you have the chance? Requires approval from a `values.yaml` maintainer

File: manifests/charts/base/crds/crd-all.gen.yaml
Comment (jacob-delgado) [inline manifests/charts/base/crds/crd-all.gen.yaml:2]: Is there a reason why this was reverted from v1 to v1beta? Maybe a bad rebase?

File: operator/cmd/mesh/install.go
Comment (jacob-delgado) [inline operator/cmd/mesh/install.go:376]: nit: consider adding a warning message

**Pull request closed without merging**
